lib/spack/docs/packaging_guide.rst

```diff
@@ -2564,98 +2564,88 @@ build system.
+package that does not support ``build_system_flags``, the build
+environment will raise a ``NotImplementedError`` .
 
-There are three primary idioms that can be combined to create whatever
-behavior the package requires.
-
-1. The default behavior for packages inheriting from
-``AutotoolsPackage`` or ``CmakePackage``.
+Here are the definitions of the three builtin flag handlers:
 
 .. code-block:: python
 
-    def default_flag_handler(self, env, flag_val):
-        env.append_flags(flag_val[0].upper(), ' '.join(flag_val[1]))
-        return []
+   def inject_flags(self, name, flags):
+       return (flags, None, None)
 
-2. The default behavior for other packages
+   def env_flags(self, name, flags):
+       return (None, flags, None)
 
-.. code-block:: python
+   def build_system_flags(self, name, flags):
+       return (None, None, flags)
 
-    def default_flag_handler(self, env, flag_val):
-        return flag_val[1]
+Note that it would be equivalent to return ``[]`` instead of ``None``.
 
-
-3. Packages may have additional flags to add to the build. These flags
-can be added to either idiom above. For example:
+Packages can override the default behavior either by specifying one of
+the builtin flag handlers
 
 .. code-block:: python
 
-    def default_flag_handler(self, env, flag_val):
-        flags = flag_val[1]
-        flags.append('-flag')
-        return flags
-
-or
-
-.. code-block:: python
+   flag_handler = <PackageClass>.env_flags
 
-    def default_flag_handler(self, env, flag_val):
-        env.append_flags(flag_val[0].upper(), ' '.join(flag_val[1]))
-        env.append_flags(flag_val[0].upper(), '-flag')
-        return []
+where ``<PackageClass>`` can be any of the subclasses of PackageBase
+discussed in :ref:`installation_procedure`.
 
-Packages may also opt for methods that include aspects of any of the
-idioms above. E.g.
+or by implementing the flag_handler method. Suppose for a package
+``Foo`` we need to pass ``cflags``, ``cxxflags``, and ``cppflags``
+through the environment, the rest of the flags through compiler
+wrapper injection, and we need to add ``-lbar`` to ``ldlibs``. The
+following flag handler method accomplishes that.
 
 .. code-block:: python
 
-    def default_flag_handler(self, env, flag_val):
-        flags = []
-        if len(flag_val[1]) > 3:
-            env.append_flags(flag_val[0].upper(), ' '.join(flag_val[1][3:]))
-            flags = flag_val[1][:3]
-        else:
-            flags = flag_val[1]
-        flags.append('-flag')
-        return flags
+   def flag_handler(self, name, flags):
+       if name in ['cflags', 'cxxflags', 'cppflags']:
+           return (None, flags, None)
+       elif name == 'ldlibs':
+           flags.append('-lbar')
+       return (flags, None, None)
 
 Because these methods can pass values through environment variables,
 it is important not to override these variables unnecessarily in other
 package methods. In the ``setup_environment`` and
```

Comment:
I would add:

> ... in other package methods when using alternative flag handlers.

(i.e. clarify when people have to worry about manually setting compiler flags).